core/src/main/java/com/linecorp/armeria/client/retry/RetryConfig.java

```diff
@@ -0,0 +1,163 @@
+    }
+
+    /**
+     * Returns config's maxTotalAttempt.
+     */
+    public int maxTotalAttempts() {
+        return maxTotalAttempts;
+    }
+
+    /**
+     * Returns config's responseTimeoutMillisForEachAttempt.
+     */
+    public long responseTimeoutMillisForEachAttempt() {
+        return responseTimeoutMillisForEachAttempt;
+    }
+
+    /**
+     * Returns config's retryRule, could be null.
+     */
+    @Nullable public RetryRule retryRule() {
+        return retryRule;
+    }
+
+    /**
+     * Returns config's retryRuleWithContent, could be null.
+     */
+    @Nullable public RetryRuleWithContent<T> retryRuleWithContent() {
+        return retryRuleWithContent;
+    }
+
+    /**
+     * Returns config's retry rule that is converted from retryRuleWithContent, could be null.
+     */
+    @Nullable public RetryRule fromRetryRuleWithContent() {
+        return fromRetryRuleWithContent;
+    }
+
+    /**
+     * Returns config's maxContentLength, which is non-zero only if a {@link RetryRuleWithContent} is used.
+     */
+    public int maxContentLength() {
+        return maxContentLength;
+    }
+
+    /**
+     * Returns whether a {@link RetryRuleWithContent} is being used.
+     */
+    public boolean needsContentInRule() {
+        return needsContentInRule;
+    }
+
+    /**
+     * Returns whether the associated requires response trailers.
+     */
+    public boolean requiresResponseTrailers() {
+        return needsContentInRule() ?
+               retryRuleWithContent().requiresResponseTrailers() : retryRule().requiresResponseTrailers();
+    }
+
+    RetryConfig(RetryRule retryRule, int maxTotalAttempts, long responseTimeoutMillisForEachAttempt) {
+        checkArguments(maxTotalAttempts, responseTimeoutMillisForEachAttempt);
+        this.retryRule = checkNotNull(retryRule);
+        this.maxTotalAttempts = maxTotalAttempts;
+        this.responseTimeoutMillisForEachAttempt = responseTimeoutMillisForEachAttempt;
+        needsContentInRule = false;
+        maxContentLength = 0;
+        retryRuleWithContent = null;
+        fromRetryRuleWithContent = null;
+    }
+
+    RetryConfig(
+            RetryRuleWithContent<T> retryRuleWithContent,
+            int maxContentLength,
+            int maxTotalAttempts,
+            long responseTimeoutMillisForEachAttempt) {
+        checkArguments(maxTotalAttempts, responseTimeoutMillisForEachAttempt);
+        this.maxContentLength = maxContentLength;
+        this.retryRuleWithContent = checkNotNull(retryRuleWithContent);
+        fromRetryRuleWithContent = RetryRuleUtil.fromRetryRuleWithContent(retryRuleWithContent);
+        this.maxTotalAttempts = maxTotalAttempts;
+        this.responseTimeoutMillisForEachAttempt = responseTimeoutMillisForEachAttempt;
+        needsContentInRule = true;
+        retryRule = null;
+    }
+
+    private RetryConfig() {
+        throw new IllegalStateException("RetryConfig must have a rule.");
+    }
```

Comment:
This class already has package-private constructors. Can't we remove this?

Reply:
Done